### PR TITLE
ISPN-11797 Introduce strong typed indexing configuration

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.redhat.com/browse/ISPN-11797

- [x] Add config for index storage
- [x] Add index reader attributes
- [x] Add index writer attributes
- [x] Deprecate property k/v based config and document migration

### CODE REVIEW

**Comment by gustavonalle** [comment]
Read for review

**Comment by tristantarrant** [comment]
/me is happy. I believe I asked @Sanne for something like this 6 years ago :)

File: core/src/main/java/org/infinispan/configuration/cache/IndexMergeConfigurationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+public class IndexMergeConfigurationBuilder extends AbstractIndexingConfigurationChildBuilder
+      implements Builder<IndexMergeConfiguration>, ConfigurationBuilderInfo {
+
+   private static final String KEY_PREFIX = "hibernate.search.backend.io.merge";
+   private static final String MAX_DOCS_KEY = KEY_PREFIX + ".max_docs";
+   private static final String FACTOR_KEY = KEY_PREFIX + ".factor";
+   private static final String MIN_SIZE_KEY = KEY_PREFIX + ".min_size";
+   private static final String MAX_SIZE_KEY = KEY_PREFIX + ".max_size";
+   private static final String MAX_FORCED_SIZE_KEY = KEY_PREFIX + ".max_forced_size";
+   private static final String CALIBRATE_BY_DELETES_KEY = KEY_PREFIX + ".calibrate_by_deletes";
```
Comment (yrodiere) [inline core/src/main/java/org/infinispan/configuration/cache/IndexMergeConfigurationBuilder.java:33]: FWIW, constants are available in Hibernate Search: see `org.hibernate.search.backend.lucene.cfg.LuceneIndexSettings`.
Similarly, for backend-level settings: `org.hibernate.search.backend.lucene.cfg.LuceneBackendSettings`.

Comment (gustavonalle) [reply core/src/main/java/org/infinispan/configuration/cache/IndexMergeConfigurationBuilder.java:33]: Unfortunately I can't use them here, as ```core``` does not depend on Hibernate Search

**Comment by gustavonalle** [comment]
Pushed a change for the queue and thread pool index defaults

**Comment by tristantarrant** [comment]
@gustavonalle needs rebase

**Comment by gustavonalle** [comment]
rebased

**Review by tristantarrant — APPROVED** [review]
1 minor comment on the xsd docs. Otherwise it's fine

File: core/src/main/resources/schema/infinispan-config-12.0.xsd [hunk 2]
```diff
@@ -2296,6 +2318,171 @@
...
+        </xs:annotation>
+      </xs:enumeration>
+    </xs:restriction>
+  </xs:simpleType>
+
+  <xs:complexType name="index-reader">
+    <xs:attribute name="refresh-interval" type="xs:long" default="0">
+      <xs:annotation>
+        <xs:documentation>
+          Advanced: Defines the internal in milliseconds to reopen the index reader.
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply core/src/main/resources/schema/infinispan-config-12.0.xsd]: Fixed!

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**